pkg/codegen/python/gen.go

```diff
@@ -456,9 +461,74 @@ func (mod *modContext) genInit(exports []string) string {
 		fmt.Fprintf(w, ")\n")
 	}
 
+	// If there are resources in this module, register the module with the runtime.
+	if len(mod.resources) != 0 {
+		mod.genResourceModule(w)
+	}
+
 	return w.String()
 }
 
+// genResourceModule generates a ResourceModule definition and the code to register an instance thereof with the
+// Pulumi runtime. The generated ResourceModule supports the deserialization of resource references into fully-
+// hydrated Resource instances. If this is the root module, this function also generates a ResourcePackage
+// definition and its registration to support rehydrating providers.
+func (mod *modContext) genResourceModule(w io.Writer) {
+	contract.Assert(len(mod.resources) != 0)
+
+	fmt.Fprintf(w, "\nimport pulumi")
```

Comment:
Unfortunately, I think this will show up in intellisense/code-completion in VS Code / PyCharm when dotting into the top-level module, e.g. writing `pulumi_aws.` will show `pulumi` in intellisense, which isn't ideal, and we've done some work recently to try to avoid this kind of thing as much as possible.

The perfect solution is to use an explicit `__all__` to control exactly what is "exported". But this would involve listing everything in the `import *` we're importing for each module. So, in lieu of using `__all__`, a simpler solution for now is to just import this like `import pulumi as _pulumi`.  The leading underscore is an indication that this is "private".

Same for the `Module` and `Package` classes -- they should be `_Module` and `_Package`.

Reply:
Or even better, could we nest all this inside a "private" function? Then we're only exposing a single "private" member.

```python
def _register_module():
    import pulumi

    class Module(pulumi.runtime.ResourceModule):
        ...

    module_instance = Module()
    pulumi.runtime.register_resource_module("a", "a", module_instance)
    pulumi.runtime.register_resource_module("b", "b", module_instance)
    pulumi.runtime.register_resource_module("c", "c", module_instance)

    class Package(pulumi.runtime.ResourcePackage):
        ...

    pulumi.runtime.register_resource_package("pkg", Package())

_register_module()
```

Reply:
That's a great idea.